Mathlib/Tactic/Linter/TerminalRefineLinter.lean

```diff
@@ -0,0 +1,75 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Lean.Elab.Command
+import Lean.Linter.Util
+
+/-!
+#  The "terminal refine" linter
+
+The "terminal refine" linter flags terminal usages of `refine`.
+-/
+
+open Lean Elab
+
+namespace Mathlib.Linter
+
+/-- The terminal refine linter emits a warning on terminal usages of `refine`. -/
+register_option linter.terminalRefine : Bool := {
+  defValue := true
+  descr := "enable the terminal refine linter"
+}
+
+namespace TerminalRefine
+
+/-- `refine? stx` detects whether the input syntax `stx` is `refine` or `refine'`. -/
+def refine? : Syntax → Bool
+  | .node _ ``Lean.Parser.Tactic.refine' _ => true
+  | .node _ ``Lean.Parser.Tactic.refine _  => true
+  | _ => false
+
+/-- `SyntaxNodeKinds` that "contain" a `refine` and that the linter should ignore. -/
+abbrev ignore : HashSet SyntaxNodeKind := HashSet.empty
+  |>.insert `Mathlib.Tactic.useSyntax
+  |>.insert `Mathlib.Tactic.«tacticUse!___,,»
+  |>.insert `Mathlib.Tactic.congrM
+
+/-- `refine_tree t` returns all terminal usages of `refine/refine'` in the input infotree. -/
+partial
+def refine_tree : InfoTree → Array Syntax
+  | .node k args =>
+    let rargs := (args.map refine_tree).toArray.flatten
+    if let .ofTacticInfo i := k then
+      let stx := i.stx
+      if ! ignore.contains stx.getKind then
+        if (refine? stx) &&
+          (i.goalsAfter.length + 1 == i.goalsBefore.length &&  -- would `<` be enough?
```

Comment:
I think that should indeed be enough, because you check on the next line that no goals where modified, right?